File: problem-of-the-day/day08/Maximum Path Sum.py
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        ans = [root.val]  # Initialize a list to store the result

        def DFS(root):
            if root == None:
                return 0  # Base case: return 0 if the node is None

            # Recursive calculation for left and right subtrees
            # Ensure that negative values are not considered (max(0, ...))
            lmax = max(0, DFS(root.left))
            rmax = max(0, DFS(root.right))

            # Update the overall maximum path sum
            ans[0] = max(ans[0] , root.val + lmax + rmax)

            # Return the maximum path sum that starts from the current node
            return root.val + max(lmax , rmax)

        # Call the DFS function on the root node
        DFS(root) 

        # Return the final maximum path sum
        return ans[0]          

File: problem-of-the-day/day08/Maximum Path Sum.cpp
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxPath = INT_MIN;  // Initialize maxPath to store the result
        DFS(root, maxPath);  // Call the DFS function
        return maxPath;  // Return the final maximum path sum
    }
private:
    int DFS(TreeNode* root, int& maxPath) {
        if (root == NULL) {
            return 0;  // Base case: return 0 if the node is NULL
        }
        
        // Recursive calculation for left and right subtrees
        // Ensure that negative values are not considered (max(DFS(...), 0))
        int lmax = max(DFS(root->left, maxPath), 0);
        int rmax = max(DFS(root->right, maxPath), 0);
        
        // Update the overall maximum path sum
        maxPath = max(maxPath, root->val + lmax + rmax);
        
        // Return the maximum path sum that starts from the current node
        return root->val + max(lmax, rmax);
    }
};